src/index.ts

```diff
@@ -0,0 +1,169 @@
+import {
+  requestsCount,
+  requestDuration,
+  requestStageDuration,
+} from "./metrics";
+import { parseUrlTemplate } from "./utils";
+
+interface Headers {
+  [header: string]: number | string | string[] | undefined;
+}
+
+interface Options {
+  baseUrl: string;
+  headers: Headers;
+}
+
+interface Timings {
+  start?: number;
+  socket?: number;
+  lookup?: number;
+  connect?: number;
+  upload?: number;
+  response?: number;
+  end?: number;
+  error?: number;
+  phases: {
+    wait?: number;
+    dns?: number;
+    tcp?: number;
+    request?: number;
+    firstByte?: number;
+    download?: number;
+    total?: number;
+  };
+}
+
+interface Result {
+  timings?: Timings;
+  statusCode?: number;
+  body?: unknown;
+}
+
+type Method =
+  | "GET"
+  | "PUT"
+  | "POST"
+  | "PATCH"
+  | "DELETE"
+  | "HEAD"
+  | "OPTIONS";
+
+export class HTTPClient {
+  baseUrl: string;
+  headers: Headers;
+
+  constructor(opts: Options) {
+    const { baseUrl, headers } = opts;
+    this.baseUrl = baseUrl;
+    this.headers = headers;
+  }
+
+  async request(
+    method: Method,
+    pathTemplate: string,
+    params = {},
+    body?: unknown,
+    options?: RequestInit
+  ): Promise<Result> {
+    const path = parseUrlTemplate(pathTemplate);
+    const url = new URL(path.expand(params), this.baseUrl);
+
+    const startTime = performance.now();
+
+    const defaultOptions: RequestInit = {
+      method: method.toUpperCase(),
+      headers: {
+        "Content-Type": "application/json",
+        ...this.headers,
+      },
+      body: body ? JSON.stringify(body) : undefined,
+    };
+
+    const fetchOptions = { ...defaultOptions, ...options };
+
+    try {
+      const response = await fetch(url.toString(), fetchOptions);
+      const endTime = performance.now();
+      if (!response.ok) {
+        const errorBody = await response.text();
+        throw new Error(
+          errorBody || response.statusText || "Request failed"
+        );
```

Comment:
This will result in errors with potentially large HTML pages as the message.

The response body will typically be a JSON payload, but it would fall back to text if it cannot be parsed.

You want a richer error type here.

v1 used to provide a GotError which had `err.statusCode` and `err.body` if available.

If we want to break as little as possible then we should make the same fields.

Reply:
I just had a look at some other projects and the fields that are currently used are

```
err.statusCode (number)
err.statusMessage (string)
err.response.body (any)
```

There could be more.

Probably need to consider if we want to be breaking here in the format, but we should have at least that data available.